src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -260,18 +263,72 @@ int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t
                     gavl_left_rotation(cur_node, rnode);
             }
 
-            if (!STACK_EMPTY(node_stack)) {
-                STACK_POP(node_stack, cur_node);
-                goto stack_recovery;
+            if (!STACK_EMPTY(gavl_tree_iptr->node_stack)) {
+                STACK_POP(gavl_tree_iptr->node_stack, cur_node);
+                continue;
             } else {
                 break;
             }
         } while (1);
 
-        while (gavl_tree_iptr->root->parent != NULL)
+        while (gavl_tree_iptr->root && gavl_tree_iptr->root->parent != NULL)
             gavl_tree_iptr->root = gavl_tree_iptr->root->parent;
     }
 
+    return;
+}
+
+int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t len,
+                         const void *val)
+{
+    int mpl_err = MPL_SUCCESS;
+    gavl_tree_node_s *node_ptr;
+    gavl_tree_s *gavl_tree_iptr = (gavl_tree_s *) gavl_tree;
+
+    node_ptr = (gavl_tree_node_s *) MPL_malloc(sizeof(gavl_tree_node_s), MPL_MEM_OTHER);
+    if (node_ptr == NULL) {
+        mpl_err = MPL_ERR_SHM_NOMEM;
+        goto fn_fail;
+    }
+    node_ptr->parent = NULL;
+    node_ptr->left = NULL;
+    node_ptr->right = NULL;
+    node_ptr->height = 1;
+    node_ptr->addr = (uintptr_t) addr;
+    node_ptr->len = len;
+    node_ptr->val = val;
+
+    if (gavl_tree_iptr->root == NULL) {
+        gavl_tree_iptr->root = node_ptr;
+    } else {
+        CLEAR_STACK(gavl_tree_iptr->node_stack);
```

Comment:
At this function, a higher level function, `STACK` is internal details that is best hidden. How about use a macro (or inline function) e.g. `GAVL_SEARCH_START`?

Reply:
I think I need to move it to lower level. I should not have any stack operation in high level.